client/allocrunner/alloc_runner.go

```diff
@@ -1233,6 +1233,84 @@ func (ar *allocRunner) Signal(taskName, signal string) error {
 	return err.ErrorOrNil()
 }
 
+// Reconnect synchronizes client state between an incoming unknown allocation and
+// the current allocRunner and appends a reconnect event to each task if alloc is running.
+func (ar *allocRunner) Reconnect(update *structs.Allocation) {
+	// Guard against the server sending update requests faster than the state
+	// gets propagated back to them.
+	if ar.hasPendingReconnect() {
+		ar.logger.Trace("skipping due to recent reconnect")
+		return
+	}
+
+	ar.maybeAppendReconnectEvent()
+
+	// Populate the alloc with updated events.
+	clientAlloc := ar.clientAlloc(ar.TaskRunnerTaskStates())
+
+	// Sync the incoming server state with the actual client state including the newly minted event.
+	update.ClientStatus = clientAlloc.ClientStatus
+	update.ClientDescription = clientAlloc.ClientDescription
+	update.TaskStates = clientAlloc.TaskStates
+
+	// Update the local copy of alloc.
+	if err := ar.stateDB.PutAllocation(update); err != nil {
+		ar.logger.Error("error reconnecting alloc when persisting updated alloc locally", "error", err, "alloc_id", update.ID)
+		return
+	}
+
+	// Sync the alloc runner. Important for ensuring the AllocModifyIndex is propagated.
+	// Do not call setAlloc on TaskRunners. That will restart the alloc.
+	ar.setAlloc(update)
+
+	// Update the server.
+	ar.stateUpdater.AllocStateUpdated(update)
+}
+
+// hasPendingReconnect checks the current alloc TaskRunnerTaskStates to see if a recent
+// reconnect event was appended. This is useful for allowing the client state to
+// have time to propagate to the servers before attempting another reconnect.
+func (ar *allocRunner) hasPendingReconnect() bool {
+	now := time.Now()
+
+	for _, taskState := range ar.TaskRunnerTaskStates() {
```

Comment:
Correct me if I'm wrong but I doubt this is going to get inlined in a way that flattens the execution of the loops. So this seems like it saves a single line of code in this function in exchange for adding a new 7 line function, looping over the task states twice instead of once, and allocating a map that we immediately throw away.

It looks like the goal is to abstract over the map that we hand to `getClientStatus` in `maybeAppendReconnectEvent`. But that also means we're calling `ar.TaskRunnerTaskStates()` twice! And if we hadn't removed `AllocState` (see my comment below on that), then the `ar.state.TaskStates` would already have this map.

Reply:
I suspect that problem will go away if you use the memoized `AllocState` because then you can check that the state has transitioned here and not just `!structs.AllocClientStatusRunning`. Take a pass through that first though and we can review again.